### ISSUE TITLE
Dynamic port for webpacker-dev-server

### ISSUE THREAD

**Issue by monai** [rails/webpacker#767]
Webpacker configuration file is [extensively reloaded](https://github.com/rails/webpacker/blob/master/lib/webpacker/railtie.rb#L38) and it seems there is no straight forward way to dynamically set dev server port, e.g., using Foreman to start Rails and webpacker-dev-server. Any suggestions how to do that?

**Comment by gauravtiwari** [comment]
What do you mean dynamic port? You want to specify a port on runtime i.e. inside your Procfile? You can change port inside `config/webpacker.yml`, no?

**Comment by monai** [comment]
I don't want to have dev server port hardcoded to `config/webpacker.yml`. I can pass dev server port via command line argument but Rails uses the port from `config/webpacker.yml`.

**Comment by javan** [comment]
> I don't want to have dev server port hardcoded to `config/webpacker.yml`

Can you expand on that? It it adding friction to your development setup in some way?

**Comment by gauravtiwari** [comment]
@javan We allow passing CLI arguments to binstubs, which can change host and port at runtime but since we also got proxy that's using values from webpacker.yml, CLI arguments are never used: https://github.com/rails/webpacker/blob/master/lib/install/bin/webpack-dev-server.tt#L29

One solution is we set explicit env vars based on these arguments so we can use them as well inside dev server configuration: https://github.com/rails/webpacker/blob/master/lib/webpacker/dev_server.rb#L28

```rb
# inside binstubs: 
ENV['WEBPACK_DEV_SERVER_PORT'] ||= args('--port')
```

```rb
ENV['WEBPACK_DEV_SERVER_PORT'] || fetch(:port)
```

**Comment by javan** [comment]
> We allow passing CLI arguments to binstubs, which can change host and port at runtime 

I still don't understand why we do that. Why can't the sole source of configuration be `webpacker.yml`?

**Comment by gauravtiwari** [comment]
Personally, I haven't really had any use for this  :) but I guess for teams where folks are using different setup, say someone using a Docker container and someone using puma-dev, it would be nice to override the settings at runtime instead of committing conflicting `config/webpacker.yml`

**Comment by javan** [comment]
I'd prefer that go in `Webpacker` itself and then remove all the CLI arg wrangling in `bin/webpack-dev-server`:
```diff
--- a/lib/webpacker/dev_server.rb
+++ b/lib/webpacker/dev_server.rb
@@ -25,7 +25,7 @@ class Webpacker::DevServer
   end
 
   def port
-    fetch(:port)
+    ENV['WEBPACK_DEV_SERVER_PORT'] || fetch(:port)
   end
```

**Comment by gauravtiwari** [comment]
Yeah, makes sense 👍

**Comment by monai** [comment]
@javan I and my coworkers usually work with at least two running Rails projects. For each project, there is separate dev server. So with this development setup, we have to hardcode unique port in `webpacker.yml` in every project. The latter snippet you posted solves our issue. Are you going to implement it or would you rather prefer PR?

**Comment by gauravtiwari** [comment]
Fixed in #843

**Comment by gauravtiwari** [comment]
3.0.2 released 👍
